pkg/services/crd/fetcher.go

```diff
@@ -0,0 +1,111 @@
+package crd
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/weaveworks/weave-gitops/core/clustersmngr"
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+const watchCRDsFrequency = 30 * time.Second
+
+type Fetcher interface {
+	IsAvailable(clusterName, name string) bool
+	IsAvailableOnClusters(name string) map[string]bool
+	UpdateCRDList()
+}
+
+func NewFetcher(ctx context.Context, logger logr.Logger, clustersManager clustersmngr.ClustersManager) Fetcher {
+	fetcher := &defaultFetcher{
+		logger:          logger,
+		clustersManager: clustersManager,
+		crds:            map[string][]v1.CustomResourceDefinition{},
+	}
+
+	go fetcher.watchCRDs(ctx)
+
+	return fetcher
+}
+
+type defaultFetcher struct {
+	sync.RWMutex
+	logger          logr.Logger
+	clustersManager clustersmngr.ClustersManager
+	crds            map[string][]v1.CustomResourceDefinition
+}
+
+func (s *defaultFetcher) watchCRDs(ctx context.Context) {
+	_ = wait.PollImmediateInfinite(watchCRDsFrequency, func() (bool, error) {
+		s.UpdateCRDList()
+
+		return false, nil
+	})
+}
+
+func (s *defaultFetcher) UpdateCRDList() {
+	s.Lock()
+	defer s.Unlock()
+
+	ctx := context.Background()
+
+	client, err := s.clustersManager.GetServerClient(ctx)
+	if err != nil {
+		s.logger.Error(err, "unable to get client pool")
+
+		return
+	}
+
+	for clusterName, client := range client.ClientsPool().Clients() {
+		crdList := &v1.CustomResourceDefinitionList{}
+
+		s.crds[clusterName] = []v1.CustomResourceDefinition{}
+
+		err := client.List(ctx, crdList)
+		if err != nil {
+			s.logger.Error(err, "unable to list crds", "cluster", clusterName)
+
+			continue
+		}
+
+		s.crds[clusterName] = crdList.Items
+	}
+}
+
+func (s *defaultFetcher) IsAvailable(clusterName, name string) bool {
+	s.Lock()
+	defer s.Unlock()
+
+	for _, crd := range s.crds[clusterName] {
+		if crd.Name == name {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (s *defaultFetcher) IsAvailableOnClusters(name string) map[string]bool {
+	result := map[string]bool{}
+
+	s.Lock()
+	defer s.Unlock()
+
+	for clusterName, crds := range s.crds {
+		// Set this to be sure the key is there with false value if the following
+		// look-up does not say it's there.
+		result[clusterName] = false
+
+		for _, crd := range crds {
+			if crd.Name == name {
```

Comment:
Usually in Kubernetes you would check the GVK or GK and not just the name. In this particular case you'd additionally query (and check) `.spec.group` and optionally `.spec.versions.name`.

Reply:
This might be a helpful resource: https://helm.sh/docs/chart_template_guide/function_list/#capabilitiesapiversionshas

Reply:
The only purpose of this endpoint is to check if given resource is available or not for example `canaries.flagger.app` to determine so we can show a getting started guide to the user if it's not installed.

Reply:
Technically this is just a port (copy + update) of the same functionality from progressive delivery, but as we go forward we want more resources to check.